torch_xla/csrc/runtime/pjrt_computation_client.cc

```diff
@@ -675,34 +664,62 @@ PjRtComputationClient::ExecuteReplicated(
                   .value();
   }
 
-  std::vector<std::vector<ComputationClient::DataPtr>> data_handles;
-  data_handles.reserve(results.size());
-  std::vector<size_t> dims(results.size());
+  size_t num_outputs = results[0].size();
+  std::vector<ComputationClient::DataPtr> data_handles(num_outputs);
 
   {
     tsl::profiler::TraceMe activity(
         "PjRtComputationClient::ExecuteReplicated_result_handle",
         tsl::profiler::TraceMeLevel::kInfo);
-    for (int32_t i = 0; i < results.size(); ++i) {
-      xla::PjRtDevice* pjrt_device = StringToPjRtDevice(devices[i]);
-      XLA_CHECK(pjrt_device->IsAddressable())
-          << pjrt_device->DebugString() << " is not addressable.";
-
-      std::vector<ComputationClient::DataPtr> datas;
-      datas.reserve(results[i].size());
-      dims[i] = results[i].size();
-      for (int32_t j = 0; j < results[i].size(); ++j) {
-        std::unique_ptr<xla::PjRtBuffer> buffer = std::move(results[i][j]);
-        XLA_CHECK(pjrt_device == buffer->device())
-            << "Exepcted device: " << pjrt_device->DebugString()
-            << " vs. actual device: " << buffer->device()->DebugString();
-
-        std::shared_ptr<PjRtData> data =
-            std::make_shared<PjRtData>(devices[i], std::move(buffer));
-        datas.push_back(data);
+
+    xla::HloModuleConfig hlo_config(computation.program_shape());
```

Comment:
This looks like we are pushing down the execution results post-processing logics from the PyTorch/XLA (sharding util) down to the runtime client.. which I am still not sure of.

Reply:
None of the substantive logic is moving down. The input and output shardings are still entirely determined by the user's sharding annotations and the compiler. The output shardings here even come from the same place they were coming from before: the compiled executable

https://github.com/pytorch/xla/blob/b20a082a7c616f9f5d08fa457d7ee4db0a03b52f/torch_xla/csrc/xla_sharding_util.cpp#L660-L670

This is just shifting the plumbing around so that `ExecuteReplicated` can consume and emit sharded data. The only reason I'm building an HloModule in this draft is because calling `GetOutputShardings`  directly on the executable is just too slow in practice. This is otherwise just the `XlaComputation` that we extract from the executable in `Compile`. 

It turns out that the shapes/shardings that we set here and in `OutputHandler` are not used, because we only use the shape/sharding on the placeholder anyway. See my other comment: https://github.com/pytorch/xla/pull/5737#discussion_r1379428346